code/obj/item/gun/energy.dm

```diff
@@ -214,26 +216,185 @@ TYPEINFO(/obj/item/gun/energy)
+		src.projectiles = null
+
+	examine()
+		. = ..()
+		if(src.evaluate_quality())
+			. += "It has had some parts replaced and is now a functional weapon. "
+		else
+			. += "While currently non-functional, it could be brought to working order with some replacement parts. "
+		if(src.panelOpen)
+			. += "Its maintenance panel is open. "
+			if(src.myLens)
+				. += "It contains a [src.myLens]. "
+			else
+				. += "It is missing a lens. "
+			if(src.myCoil)
+				. += "It contains a [src.myCoil]. "
+			else
+				. += "It is missing a small coil. "
+
+	attackby(obj/item/object, mob/user)
+		. = ..()
+		if(isscrewingtool(object))
+			user.show_text(SPAN_NOTICE("You [src.panelOpen ? "close" : "open"] the maintenance panel."))
+			playsound(user, 'sound/items/Screwdriver2.ogg', 65, TRUE)
+			src.panelOpen = !src.panelOpen
+			if(!src.panelOpen && src.evaluate_quality())
+				if(src.customProjectileDamage == CAPTAINGUN_T3_DAMAGE && src.customProjectileCost == CAPTAINGUN_T3_COST)
+					user.unlock_medal("Tinkerer", 1)
+		if(istype(object, /obj/item/coil/small))
+			if(src.panelOpen)
+				user.show_text(SPAN_NOTICE("You insert the [object] into the [src]."))
+				playsound(user, 'sound/items/Deconstruct.ogg', 65, TRUE)
+				user.put_in_hand_or_eject(src.myCoil)
+				object.set_loc(src)
+				user.u_equip(object)
+				src.myCoil = object
+			else
+				user.show_text(SPAN_NOTICE("The [src]'s maintenance panel is closed."))
+		if(istype(object, /obj/item/lens))
+			if(src.panelOpen)
+				user.show_text(SPAN_NOTICE("You insert the [object] into the [src]."))
+				playsound(user, 'sound/items/Deconstruct.ogg', 65, TRUE)
+				user.put_in_hand_or_eject(src.myLens)
+				object.set_loc(src)
+				user.u_equip(object)
+				src.myLens = object
+			else
+				user.show_text(SPAN_NOTICE("The [src]'s maintenance panel is closed."))
+
+	canshoot(mob/user)
+		if(!src.evaluate_quality())
+			return FALSE
+		//change the projectile here because this is checked by both gun.shoot_point_blank(), gun.shoot(), and gun.suicide()
+		//which duplicate a horrifying amount of code
+		src.set_current_projectile(new /datum/projectile/laser/custom(src.customProjectileDamage, src.customProjectileCost))
+		. = ..()
+
+	//case-in-point, couldnt avoid it here because I needed it to happen after the shot
+	//and because canshoot() is sometimes checked when we arent actually firing a shot and we dont want to blow up randomly
+	shoot(turf/target, turf/start, mob/user, POX, POY, is_dual_wield, atom/called_target)
+		. = ..()
+		if(.)
+			//exactly enough that an erebite component will have a chance to explode from a normal damage laser
+			//and guaranteed to explode from a high damage laser
+			//also melts ice so it isnt viable for as a normal damage lens material because thats too easy
+			src.myLens.material_trigger_on_temp(FROM_CELSIUS(src.current_projectile.power * 20))
+			src.myCoil.material_trigger_on_temp(FROM_CELSIUS(src.current_projectile.power * 20))
```

Comment:
These should be using `temperature_expose` instead of directly calling the material triggers.

Reply:
`temperature_expose(datum/gas_mixture/air, exposed_temperature, exposed_volume)` is called by airgroups on the atoms they are touching. `material_trigger_on_temp(var/temperature_applied)` is the appropriate proc for some other process exposing a thing to temperature.